keps/sig-apps/20190326-stateful-application-api.md

```diff
@@ -0,0 +1,790 @@
+type ApplicationSnapshotContentSpec struct {
+	// SnapshotClassName is the name of the ApplicationSnapshotClass object to use for snapshot-related operations.
+	SnapshotClassName string `json:"snapshotClassName"`
+
+	// SnapshotRef is an reference to the ApplicationSnapshot resource this ApplicationSnapshotContent resource is
+	// bound to. When specified, this creates a bi-directional binding between an ApplicationSnapshot resource and
+	// this ApplicationSnapshotContent resource.
+	// +optional
+	SnapshotRef *corev1.ObjectReference `json:"snapshotRef,omitempty"`
+
+	// VolumeSnapshotRefs is mapping from PVC names to references to the corresponding VolumeSnapshot objects created
+	// for taking snapshots of individual persistent volumes mounted by the application when the snapshot source is
+	// a StatefulApplication.
+	// +optional
+	VolumeSnapshotRefs map[string]corev1.ObjectReference `json:"volumeSnapshotRefs,omitempty"`
+
+	// Tags are user-specified free texts to attach to the snapshot. This combines user-specified tags in the
+	// ApplicationSnapshot and tags returned by pre-hooks.
+	// +optional
+	Tags map[string]string `json:"tags,omitempty"`
+
+	// MetadataHandle is the handle to persisted application metadata. It contains information needed to read the
+	// metadata back.
+	// +optional
+	MetadataHandle *MetadataHandle `json:"metadataHandle,omitempty"`
+
+	// DeletionPolicy is the policy on what happens to an ApplicationSnapshotContent resource and the
+	// local application snapshot it represents when the ApplicationSnapshot resource it is bound to
+	// is deleted.
+	// DeletionPolicy set here takes precedence over the one (if set) in ApplicationSnapshotClass, if set.
+	// +optional
+	// +kubebuilder:validation:Enum=Delete,Retain
+	DeletionPolicy *DeletionPolicy `json:"deletionPolicy,omitempty"`
+}
+
+type ApplicationSnapshotContentStatus struct {}
+
+// MetadataHandleType is the type of the application metadata handle.
+type MetadataHandleType string
+
+// Set of valid MetadataHandleType values.
+const (
+	MetadataHandleOnDisk = "OnDisk"
+)
+
+// MetadataHandle is a handle to persisted application metadata, which includes
+// specs of Kubernetes resources that make up an application.
+// TODO: complete this when extending the storage options of application metadata.
+type MetadataHandle struct {
+	// Type is the type of the MetadataHandle.
+	Type MetadataHandleType `json:"type"`
+
+	// OnDisk is a type of application metadata handles that store metadata on disk.
+	OnDisk *OnDiskHandle `json:"onDisk,omitempty"`
+}
+
+// OnDiskHandle is a type of application metadata handles that store metadata on disk.
+type OnDiskHandle struct {
+	Directory string `json:"directory"`
+}
+
+type ApplicationSnapshotContent struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ApplicationSnapshotContentSpec   `json:"spec,omitempty"`
+	Status ApplicationSnapshotContentStatus `json:"status,omitempty"`
```

Comment:
Will this status be different from ApplicationSnapshotStatus?

Reply:
I deleted this status given that it's empty and not really used. There's also no foreseeable use of this in the future.